model/ftn/ww3_prnc.ftn

```diff
@@ -1804,23 +1802,27 @@
 !
                 DO IY=1,NY
                   DO IX=1,NX
-                    FA(IX,IY)                                         &
-                          = RD11(IX,IY) * XC(IX21(IX,IY),IY21(IX,IY)) &
-                          + RD21(IX,IY) * XC(IX22(IX,IY),IY21(IX,IY)) &
-                          + RD12(IX,IY) * XC(IX21(IX,IY),IY22(IX,IY)) &
-                          + RD22(IX,IY) * XC(IX22(IX,IY),IY22(IX,IY))
+                    CALL POINT_INTERP(XC(IX21(IX,IY),IY21(IX,IY)),     &
+                               XC(IX22(IX,IY),IY21(IX,IY)),            &
+                               XC(IX21(IX,IY),IY22(IX,IY)),            &
+                               XC(IX22(IX,IY),IY22(IX,IY)),            &
+                               XC(IX,IY), FILLVALUE, RD11(IX,IY),      &
+                               RD21(IX,IY), RD12(IX,IY), RD22(IX,IY),  &
+                               FA(IX,IY))
                     END DO
                   END DO
 !
                 IF (NFCOMP.EQ.2) THEN
 !/O3             IF ( IAPROC .EQ. NAPOUT )    WRITE (NDSO,976) ' (2) '
                     DO IY=1,NY
```

Comment:
This looks like a better approach, Juan.
My only concern is that the `POINT_INTERP` subroutine is being called for every grid cell. This feels like an unnecessary overhead popping stuff on and off the call stack. Can `POINT_INTERP` take arrays instead of scalars and loop over `NX` and `NY` internally?

Reply:
Thanks for the comment, I thought following this approach, and I was not sure what was more efficient. In any case I guess the loss of performance will not be very important if any, but the code will be clearer too. I will implement the changes and commit when they are ready.

Reply:
The changes are now implemented. I tested them and they do not give exactly the same answer than the previous version of the code, but the differences are so small that can be considered rounding errors.

Reply:
That's odd - I wouldn't have expected moving the loop into the subroutine to have changed the values at all...Perhaps it is an effect of compiler optimization?

Reply:
You are right, in the previous version of the code there was a mismatch between the parameters of the subroutine and the calls to it: RD21 and RD12 were exchanged, so the code was wrong. I believe the code is now right and reproduces the original behaviour, please double check.

Reply:
Great. I can't see in the code where the error was (with respect to RD21 and RD22 being in the incorrect order), but if it is now giving consistent results with the version prior to you doing the refactoring then I am happy it is good.

Reply:
Thanks Chris. The results before and after refactoring are the same for the density field (I was centred in fixing this), but there are differences in the wind field, and the code before refactoring seems to provide a better interpolation. I will investigate this point and modify the code until I am satisfied with the results.

Reply:
The code is now fixed. What was missing was calling the new interpolation routine with the right fill value for derived variables. The results of this new code are exactly the same as the code before refactoring.

Reply:
Nice one @ukmo-juan-castillo